resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/AutoBuildJob.java

```diff
@@ -99,6 +99,7 @@ synchronized void build(boolean needsBuild) {
 			case Job.SLEEPING :
 				wakeUp(delay);
 				break;
+			case RUNNING :
```

Comment:
or even better: call schedule() independent of job state, as the job state is not synchronized and may change in parallel.

Reply:
think I have found a caveat.

The originally missing RUNNING case did prevent a constant rescheduling of the AutoBuildJob.
doBuild() ends in a call to Workspace.endOperation() -> BuildManager.endTopLevel(). 
The latter will call build(false). And if case RUNNABLE schedules the build. -> endless loop.

So I think I need to distinguish cases and only schedule the build if `state != RUNNING || buildNeeded`.

What about the SLEEPING case? Should I even schedule in that case or leave the wakeUp()? WDYT?

Will update the PR tomorrow...

Reply:
Is the current change fixing your issue? If yes, I would consider not touch the code anymore, as long as you don't have a case to verify there is still some problem.

Reply:
it should only schedule if needsBuild

Reply:
@iloveeclipse the problem is that I need to run our UITest 10-20 times repeatedly to reproduce the issue. This makes it difficult to tell whether the race is completely fixed or not. But I think I'll invest the additional time and check if I can run this over the weekend to get it stable with a minimal change.

Reply:
if  not shouldBuild() it should not even prepare/begin/endOperation but should exit early as the event was already consumed

Reply:
@jukzi I am a bit worried about regressions. 
This code is so old and untouched for years and I fear no one is left who knows why it seems so complex :-(
For example, there is this comment in line [106](https://github.com/eclipse-platform/eclipse.platform/blob/master/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/AutoBuildJob.java#L106) about sending events even if the autobuild is off. 
If we add an early exit where before there was none couldn’t that also have unwanted side-effects?

Reply:
After thinking about this once more on Friday, I decided to go for the least intrusive solution:
The scheduling of the AutoBuildJob is unchanged in all states except RUNNING.
In the RUNNING state, it is rescheduled, if autobuild is on and buildNeeded is true.